pkg/placement/raft/fsm.go

```diff
@@ -0,0 +1,122 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package raft
+
+import (
+	"io"
+	"sync"
+
+	"github.com/hashicorp/go-msgpack/codec"
+	"github.com/hashicorp/raft"
+)
+
+// CommandType is the type of raft command in log entry
+type CommandType uint8
+
+const (
+	// MemberUpsert is the command to update or insert new or existing member info
+	MemberUpsert CommandType = 0
+	// MemberRemove is the command to remove member from actor host member state
+	MemberRemove = 1
+)
+
+// FSM implements a finite state machine that is used
+// along with Raft to provide strong consistency. We implement
+// this outside the Server to avoid exposing this outside the package.
+type FSM struct {
+	// stateLock is only used to protect outside callers to State() from
+	// racing with Restore(), which is called by Raft (it puts in a totally
+	// new state store). Everything internal here is synchronized by the
+	// Raft side, so doesn't need to lock this.
+	stateLock sync.RWMutex
+	state     *DaprHostMemberState
+}
+
+func newFSM() *FSM {
+	return &FSM{
+		state: newDaprHostMemberState(),
+	}
+}
+
+// State is used to return a handle to the current state
+func (c *FSM) State() *DaprHostMemberState {
+	c.stateLock.RLock()
+	defer c.stateLock.RUnlock()
+	return c.state
+}
+
+func (c *FSM) upsertMember(cmdData []byte) error {
+	c.stateLock.Lock()
+	defer c.stateLock.Unlock()
+
+	var host DaprHostMember
+	if err := unmarshalMsgPack(cmdData, &host); err != nil {
+		return err
+	}
+
+	return c.state.upsertMember(&host)
+}
+
+func (c *FSM) removeMember(cmdData []byte) error {
```

Comment:
Can't the operations be done as a list? This way, placement can be less chatty by sending multiple membership changes in a single call.

Reply:
we can consider using [applyBatch](https://github.com/hashicorp/raft/blob/master/fsm.go#L37) later. this is the initial pr, we can decide it after running test later.

Reply:
This should be done before v1.0.